src/library/EndpointManagement/EndpointManagementClient.cs

```diff
@@ -494,6 +454,84 @@ private async Task EnsureEndpointManagerRunningAsync(CancellationToken cancellat
             }
         }
 
+        public Dictionary<string, string> GetEnvironmentVariables()
+        {
+            if (!string.IsNullOrEmpty(_environmentVariables.DotNetRoot))
+            {
+                _log.Info("Setting the '{0}' environment variable with '{1}' while launching '{2}'.", EnvironmentVariables.Names.DotNetRoot, _environmentVariables.DotNetRoot, EndpointManager.ProcessName);
+                return new Dictionary<string, string>() { { EnvironmentVariables.Names.DotNetRoot, _environmentVariables.DotNetRoot } };
+            }
+
+            return null;
+        }
+
+        public async Task<(string fileName, string command)> GetEndpointManagerLaunchArguments(string currentUserName, string logFileDirectory, CancellationToken cancellationToken)
+        {
+            var fileName = string.Empty;

```

Comment:
I think it'd be preferable to declare `fileName` and `command` _and return them_ in each branch of the `_platform` switch. This avoids bugs arising from forgetting a `break` or forgetting to assign them in one of the branches. I notice they don't get set in the `platform.IsWindows` branch, and I wonder if they should be? If not, then explicitly returning `("", "")` from that branch would make that decision clearer.

Reply:
Thinking about this some more, I'm wondering if it'd make more sense to have completely separate code paths for the different platforms. One of the key benefits of performing the conditional logic inside the function is that the caller shouldn't need to care which platform it is. But in this case, the caller does need to know that, because only under certain circumstances does it make sense to swap `pkexec` with `sudo`.

I'm thinking along the lines of separate classes for each platform, and the Linux one would internally do the retry with `sudo`.

This would also address another slight concern I had, which is that this function is doing a lot more stuff in the Windows branch, i.e. it's launching a process _and_ waiting for the endpoint manager to become ready.

Reply:
updated, thank for the suggestion @peterbom